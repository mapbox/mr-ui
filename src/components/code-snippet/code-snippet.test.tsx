import React from 'react';
import hljs from 'highlight.js';
import { screen, render, fireEvent } from '@testing-library/react';
import CodeSnippet from './code-snippet';

const snippetA = `dependencies {
  compile('com.mapbox.mapboxsdk:mapbox-android-sdk:5.0.2@aar') {
    transitive=true
  }
}`;

const snippetB = `import Mapbox

  class ViewController: UIViewController {
  override func viewDidLoad() {
    super.viewDidLoad()

    let url = URL(string: "mapbox://styles/mapbox/streets-v9")
    let mapView = MGLMapView(frame: view.bounds, styleURL: url)
    mapView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
    mapView.setCenter(CLLocationCoordinate2D(latitude: 59.31, longitude: 18.06), zoomLevel: 9, animated: false)
    view.addSubview(mapView)
  }
}`;

describe('CodeSnippet', () => {
  // Almost all of the logic in CodeSnippet is about determining element rendering,
  // and snapshots cover that
  describe('non highlighted', () => {
    const props = {
      code: snippetA
    } as const;

    test('renders as expected', () => {
      const { baseElement } = render(<CodeSnippet {...props} />)
      expect(baseElement).toMatchSnapshot();
    });
  });

  describe('highlighted', () => {
    const props = {
      code: snippetB,
      highlightedCode: hljs.highlightAuto(snippetA).value
    } as const;

    test('renders as expected', () => {
      const { baseElement } = render(<CodeSnippet {...props} />)
      expect(baseElement).toMatchSnapshot();
    });
  });

  describe('highlighted, no ranges, with onCopy callback', () => {
    const onCopyMock = jest.fn();
    const props = {
      code: snippetA,
      highlightedCode: hljs.highlightAuto(snippetA).value,
      onCopy: onCopyMock
    } as const;

    test('renders as expected', () => {
      const { baseElement } = render(<CodeSnippet {...props} />)
      expect(baseElement).toMatchSnapshot();
    });

    test('onCopy works', () => {
      render(<CodeSnippet {...props} />)
      fireEvent.click(screen.getByTestId('copy-button'));
      expect(onCopyMock).toHaveBeenCalledTimes(1);

      /*
      await waitFor(() => {
        expect(screen.getByTestId('firstName-error')).toBeInTheDocument();
      });
      */
    });
  });

  describe('copy ranges', () => {
    const props = {
      code: snippetB,
      highlightedCode: hljs.highlightAuto(snippetB).value,
      copyRanges: [
        [1, 1],
        [7, 11]
      ]
    };

    test('renders as expected', () => {
      const { baseElement } = render(<CodeSnippet {...props} />)
      expect(baseElement).toMatchSnapshot();
    });
  });
});

/*
testCases.copyRangesWithCallback = {
  description: 'copy ranges, with onCopy callback',
  component: CodeSnippet,
  props: {
    code: snippetB,
    highlightedCode: hljs.highlightAuto(snippetB).value,
    copyRanges: [
      [1, 1],
      [7, 11]
    ],
    onCopy: jest.fn()
  }
};

  describe(testCases.copyRangesWithCallback.description, () => {
    beforeEach(() => {
      testCase = testCases.copyRangesWithCallback;
      wrapper = shallow(
        React.createElement(testCase.component, testCase.props),
        { disableLifecycleMethods: true }
      );
    });

    test('renders as expected', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('onCopy works', () => {
      wrapper.find('CopyButton').get(0).props.onCopy();
      expect(testCase.props.onCopy).toHaveBeenCalledTimes(1);
      expect(testCase.props.onCopy).toHaveBeenCalledWith(0);
    });

    test('onCopy works again, providing correct index', () => {
      wrapper.find('CopyButton').get(1).props.onCopy();
      expect(testCase.props.onCopy).toHaveBeenCalledTimes(1);
      expect(testCase.props.onCopy).toHaveBeenCalledWith(1);
    });
  });

testCases.maxHeight = {
  description: 'maximum height',
  component: CodeSnippet,
  props: {
    code: snippetB,
    highlightedCode: hljs.highlightAuto(snippetB).value,
    copyRanges: [
      [1, 1],
      [7, 11]
    ],
    maxHeight: 120
  }
};

  describe(testCases.maxHeight.description, () => {
    beforeEach(() => {
      testCase = testCases.maxHeight;
      wrapper = shallow(
        React.createElement(testCase.component, testCase.props),
        { disableLifecycleMethods: true }
      );
    });

    test('renders as expected', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('adjusts copy blocks on resize', () => {
    let mockWindow;

    beforeEach(() => {
      testCase = testCases.copyRanges;
      mockWindow = _.assign({}, window, {
        addEventListener: jest.fn(),
        removeEventListener: jest.fn()
      });
      getWindow.mockReturnValue(mockWindow);
      wrapper = mount(React.createElement(testCase.component, testCase.props));
    });

    test('adds resize listener on mount, with adjustPositions callback', () => {
      expect(mockWindow.addEventListener).toHaveBeenCalledTimes(1);
      expect(mockWindow.addEventListener).toHaveBeenCalledWith(
        'resize',
        wrapper.instance().adjustPositions
      );
    });

    test('removes resize listener on unmount', () => {
      const adjustPositions = wrapper.instance().adjustPositions;
      wrapper.unmount();
      expect(mockWindow.removeEventListener).toHaveBeenCalledTimes(1);
      expect(mockWindow.removeEventListener).toHaveBeenCalledWith(
        'resize',
        adjustPositions
      );
    });
  });
});
*/
