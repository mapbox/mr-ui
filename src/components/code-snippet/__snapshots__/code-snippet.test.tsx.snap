// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeSnippet copy ranges renders as expected 1`] = `
<body>
  <div>
    <div
      class="relative round z0 scroll-styled"
    >
      <pre>
        <code
          class="px0 hljs"
        >
          <div
            class="relative z2"
            data-chunk-code="chunk-0"
          />
          <div
            class="relative z2"
            data-chunk-code="chunk-1"
          >
            <div
              class="pr12 py3"
              style="padding-left: 12px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                import Mapbox
              </div>
            </div>
          </div>
          <div
            class="relative z2"
            data-chunk-code="chunk-2"
          >
            <div
              class="pr12 opacity75"
              style="padding-left: 12px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                 
              </div>
            </div>
            <div
              class="pr12 opacity75"
              style="padding-left: 26.45px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                <span
                  class="hljs-keyword"
                >
                  class
                </span>
                 ViewController: UIViewController {
              </div>
            </div>
            <div
              class="pr12 opacity75"
              style="padding-left: 26.45px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                override func view
                <span
                  class="hljs-constructor"
                >
                  DidLoad()
                </span>
                 {
              </div>
            </div>
            <div
              class="pr12 opacity75"
              style="padding-left: 40.9px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                super.view
                <span
                  class="hljs-constructor"
                >
                  DidLoad()
                </span>
              </div>
            </div>
            <div
              class="pr12 opacity75"
              style="padding-left: 12px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                 
              </div>
            </div>
          </div>
          <div
            class="relative z2"
            data-chunk-code="chunk-3"
          >
            <div
              class="pr12 py3"
              style="padding-left: 40.9px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                <span
                  class="hljs-keyword"
                >
                  let
                </span>
                 url = 
                <span
                  class="hljs-constructor"
                >
                  URL(
                  <span
                    class="hljs-params"
                  >
                    string
                  </span>
                  : 
                  <span
                    class="hljs-string"
                  >
                    "mapbox://styles/mapbox/streets-v9"
                  </span>
                  )
                </span>
              </div>
            </div>
            <div
              class="pr12 py3"
              style="padding-left: 40.9px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                <span
                  class="hljs-keyword"
                >
                  let
                </span>
                 mapView = 
                <span
                  class="hljs-constructor"
                >
                  MGLMapView(
                  <span
                    class="hljs-params"
                  >
                    frame
                  </span>
                  : 
                  <span
                    class="hljs-params"
                  >
                    view
                  </span>
                  .
                  <span
                    class="hljs-params"
                  >
                    bounds
                  </span>
                  , 
                  <span
                    class="hljs-params"
                  >
                    styleURL
                  </span>
                  : 
                  <span
                    class="hljs-params"
                  >
                    url
                  </span>
                  )
                </span>
              </div>
            </div>
            <div
              class="pr12 py3"
              style="padding-left: 40.9px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                mapView.autoresizingMask = 
                <span
                  class="hljs-literal"
                >
                  [.
                  <span
                    class="hljs-identifier"
                  >
                    flexibleWidth
                  </span>
                  , .
                  <span
                    class="hljs-identifier"
                  >
                    flexibleHeight
                  </span>
                  ]
                </span>
              </div>
            </div>
            <div
              class="pr12 py3"
              style="padding-left: 40.9px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                mapView.set
                <span
                  class="hljs-constructor"
                >
                  Center(CLLocationCoordinate2D(
                  <span
                    class="hljs-params"
                  >
                    latitude
                  </span>
                  : 59.31, 
                  <span
                    class="hljs-params"
                  >
                    longitude
                  </span>
                  : 18.06)
                </span>
                , zoomLevel: 
                <span
                  class="hljs-number"
                >
                  9
                </span>
                , animated: 
                <span
                  class="hljs-literal"
                >
                  false
                </span>
                )
              </div>
            </div>
            <div
              class="pr12 py3"
              style="padding-left: 40.9px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                view.add
                <span
                  class="hljs-constructor"
                >
                  Subview(
                  <span
                    class="hljs-params"
                  >
                    mapView
                  </span>
                  )
                </span>
              </div>
            </div>
          </div>
          <div
            class="relative z2"
            data-chunk-code="chunk-4"
          >
            <div
              class="pr12 opacity75"
              style="padding-left: 26.45px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                }
              </div>
            </div>
            <div
              class="pr12 opacity75"
              style="padding-left: 12px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                }
              </div>
            </div>
          </div>
        </code>
      </pre>
      <div
        class="bg-darken75 absolute left right"
        data-chunk-overlay="chunk-1"
        style="opacity: 0; transition: opacity 300ms linear;"
      >
        <div
          class="bg-blue h-full w6"
        />
      </div>
      <div
        class="bg-darken75 absolute left right"
        data-chunk-overlay="chunk-3"
        style="opacity: 0; transition: opacity 300ms linear;"
      >
        <div
          class="bg-blue h-full w6"
        />
      </div>
      <div
        class="absolute z3 right mr3 color-white"
        data-chunk-copy="chunk-1"
        style="opacity: 0; transition: opacity 300ms linear;"
      >
        <div
          aria-label="Copy"
          data-clipboard-text="import Mapbox
"
          data-testid="copy-button"
          role="button"
        >
          <div
            aria-controls="radix-4"
            aria-expanded="false"
            aria-haspopup="dialog"
            data-state="closed"
            type="button"
          >
            <button
              class="btn btn--xs py3 px3 round"
              data-state="closed"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="events-none icon"
                data-testid="icon-clipboard"
                focusable="false"
                style="width: 18px; height: 18px;"
              >
                <use
                  xlink:href="#icon-clipboard"
                  xmlns:xlink="http://www.w3.org/1999/xlink"
                />
              </svg>
              <span
                style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
              >
                clipboard
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        class="absolute z3 right mr3 color-white"
        data-chunk-copy="chunk-3"
        style="opacity: 0; transition: opacity 300ms linear;"
      >
        <div
          aria-label="Copy"
          data-clipboard-text="    let url = URL(string: \\"mapbox://styles/mapbox/streets-v9\\")
    let mapView = MGLMapView(frame: view.bounds, styleURL: url)
    mapView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
    mapView.setCenter(CLLocationCoordinate2D(latitude: 59.31, longitude: 18.06), zoomLevel: 9, animated: false)
    view.addSubview(mapView)
"
          data-testid="copy-button"
          role="button"
        >
          <div
            aria-controls="radix-5"
            aria-expanded="false"
            aria-haspopup="dialog"
            data-state="closed"
            type="button"
          >
            <button
              class="btn btn--xs py3 px3 round"
              data-state="closed"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="events-none icon"
                data-testid="icon-clipboard"
                focusable="false"
                style="width: 18px; height: 18px;"
              >
                <use
                  xlink:href="#icon-clipboard"
                  xmlns:xlink="http://www.w3.org/1999/xlink"
                />
              </svg>
              <span
                style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
              >
                clipboard
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`CodeSnippet highlighted renders as expected 1`] = `
<body>
  <div>
    <div
      class="relative round z0 scroll-styled"
    >
      <pre>
        <code
          class="px0 hljs"
        >
          <div
            class="relative z2"
            data-chunk-code="chunk-0"
          >
            <div
              class="pr12"
              style="padding-left: 12px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                <span
                  class="hljs-keyword"
                >
                  dependencies
                </span>
                 {
              </div>
            </div>
            <div
              class="pr12"
              style="padding-left: 26.45px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                <span
                  class="hljs-keyword"
                >
                  compile
                </span>
                (
                <span
                  class="hljs-string"
                >
                  'com.mapbox.mapboxsdk:mapbox-android-sdk:5.0.2@aar'
                </span>
                ) {
              </div>
            </div>
            <div
              class="pr12"
              style="padding-left: 40.9px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                transitive=
                <span
                  class="hljs-keyword"
                >
                  true
                </span>
              </div>
            </div>
            <div
              class="pr12"
              style="padding-left: 26.45px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                }
              </div>
            </div>
            <div
              class="pr12"
              style="padding-left: 12px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                }
              </div>
            </div>
          </div>
        </code>
      </pre>
    </div>
  </div>
</body>
`;

exports[`CodeSnippet highlighted, no ranges, with onCopy callback renders as expected 1`] = `
<body>
  <div>
    <div
      class="relative round z0 scroll-styled"
    >
      <pre>
        <code
          class="px0 hljs"
        >
          <div
            class="relative z2"
            data-chunk-code="chunk-0"
          >
            <div
              class="pr12"
              style="padding-left: 12px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                <span
                  class="hljs-keyword"
                >
                  dependencies
                </span>
                 {
              </div>
            </div>
            <div
              class="pr12"
              style="padding-left: 26.45px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                <span
                  class="hljs-keyword"
                >
                  compile
                </span>
                (
                <span
                  class="hljs-string"
                >
                  'com.mapbox.mapboxsdk:mapbox-android-sdk:5.0.2@aar'
                </span>
                ) {
              </div>
            </div>
            <div
              class="pr12"
              style="padding-left: 40.9px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                transitive=
                <span
                  class="hljs-keyword"
                >
                  true
                </span>
              </div>
            </div>
            <div
              class="pr12"
              style="padding-left: 26.45px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                }
              </div>
            </div>
            <div
              class="pr12"
              style="padding-left: 12px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                }
              </div>
            </div>
          </div>
        </code>
      </pre>
      <div
        class="absolute z2 top right mr6 mt6 color-white"
      >
        <div
          aria-label="Copy"
          data-clipboard-text="dependencies {
  compile('com.mapbox.mapboxsdk:mapbox-android-sdk:5.0.2@aar') {
    transitive=true
  }
}"
          data-testid="copy-button"
          role="button"
        >
          <div
            aria-controls="radix-0"
            aria-expanded="false"
            aria-haspopup="dialog"
            data-state="closed"
            type="button"
          >
            <button
              class="btn btn--xs py3 px3 round"
              data-state="closed"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="events-none icon"
                data-testid="icon-clipboard"
                focusable="false"
                style="width: 18px; height: 18px;"
              >
                <use
                  xlink:href="#icon-clipboard"
                  xmlns:xlink="http://www.w3.org/1999/xlink"
                />
              </svg>
              <span
                style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
              >
                clipboard
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`CodeSnippet non highlighted renders as expected 1`] = `
<body>
  <div>
    <div
      class="relative round z0 scroll-styled"
    >
      <pre>
        <code
          class="px0 hljs"
        >
          <div
            class="relative z2"
            data-chunk-code="chunk-0"
          >
            <div
              class="pr12"
              style="padding-left: 12px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                dependencies {
              </div>
            </div>
            <div
              class="pr12"
              style="padding-left: 26.45px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                compile('com.mapbox.mapboxsdk:mapbox-android-sdk:5.0.2@aar') {
              </div>
            </div>
            <div
              class="pr12"
              style="padding-left: 40.9px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                transitive=true
              </div>
            </div>
            <div
              class="pr12"
              style="padding-left: 26.45px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                }
              </div>
            </div>
            <div
              class="pr12"
              style="padding-left: 12px;"
            >
              <div
                style="text-indent: -14.45px; margin-left: 14.45px;"
              >
                }
              </div>
            </div>
          </div>
        </code>
      </pre>
    </div>
  </div>
</body>
`;
